src/gprs.c

```diff
@@ -3186,21 +3192,32 @@ static void gprs_load_settings(struct ofono_gprs *gprs, const char *imsi)
 static struct pri_context *gprs_context_for_ia(struct ofono_gprs *gprs)
 {
 	GSList *l;
+	struct pri_context *ctx_ia = NULL;
+	struct pri_context *ctx_inet = NULL;	
 
 	for (l = gprs->contexts; l; l = l->next) {
 		struct pri_context *ctx = l->data;
 
-		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_IA)
-			return ctx;
-	}
+		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_IA) {
+			if (ctx->preferred)
+				return ctx;
 
-	for (l = gprs->contexts; l; l = l->next) {
-		struct pri_context *ctx = l->data;
+			if (ctx_ia == NULL)
+				ctx_ia = ctx;
+		} else if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_INTERNET &&
```

Comment:
I have changed the PR so now the priority order for the IA APN is:

1. Preferred IA context
2. IA context
3. Preferred internet context
4. Internet context
5. Other context type with preferred flag set
6. Other context type

Finally, yes, it is possible that the function returns a MMS type if it is the only one defined. Should not be an issue, and it is the same behaviour as Android's.